finetune/lora.py

```diff
@@ -40,50 +44,79 @@
 weight_decay = 0.0
 max_seq_length = 256  # see scripts/prepare_alpaca.py
 lora_r = 8
-lora_alpha = 16
+lora_alpha = 8
 lora_dropout = 0.05
 warmup_iters = 100
-
+device = torch.device("cuda:7")
+
+def swap_for_qlora(model: torch.nn.Module) -> None:
+    print("Swapping for Qlora...")
+    for name, child in tqdm(model.named_modules(), total = len(list(model.named_modules()))):
+        if name.endswith("mlp"):
+            w1 = child.c_fc1.weight
+            w2 = child.c_fc2.weight
+            w3 = child.c_proj.weight
+            qlora_config = QloraConfig(lora_r, lora_alpha, lora_dropout)
+            setattr(model, name, QloraMLP(w1.to(torch.bfloat16), w2.to(torch.bfloat16), w3.to(torch.bfloat16), qlora_config))
+
+def swap_for_qlora_jank(model: torch.nn.Module) -> None:
+    print("Swapping for Qlora...")
+    def reset_qlora(module):
+        if isinstance(module, QloraLinear):
+            module.reset_parameters()
+
+    for module in tqdm(model.transformer.h):
+        current_mlp = module.mlp
+        w1 = current_mlp.c_fc1.weight
+        w2 = current_mlp.c_fc2.weight
+        w3 = current_mlp.c_proj.weight
+        qlora_config = QloraConfig(lora_r, lora_alpha, lora_dropout)
+        new_mod = QloraMLP(w1.to(torch.bfloat16), w2.to(torch.bfloat16), w3.to(torch.bfloat16), qlora_config)
+        new_mod.apply(reset_qlora)
```

Comment:
Probably missing something here but I am a bit confused about why we need to call `reset_parameters` for QLoraLinear over here -- why don't we call `reset_parameters` at the bottom of `QLoraLinear.__init__` to initialize A and B